### PR TITLE
Preview deployments

### PR DESCRIPTION
**Pull request by Aierie**
CS-388

This PR aims to provide preview sites for PRs opened targeting master.

### CODE REVIEW

**Comment by Aierie** [comment]
@lukemelia it looks like the bucket for boxel's assets doesn't exist: http://boxel-assets.cardstack.com.s3.amazonaws.com

the bucket for s3-index exists and has an appropriate-looking index.html: http://boxel-preview.cardstack.com.s3-website-us-east-1.amazonaws.com/preview-deployments/

**Comment by Aierie** [comment]
Brief description of the behaviour of the updated workflow:

1. When does this workflow run?
When someone opens up a pull request to `master`, updates a branch with an open PR to `master`, or pushes an update/merges a PR into `master`, we run the `test-and-deploy` workflow.
https://github.com/cardstack/boxel/blob/fe7b727ea0b2952eb364537b2236a2a8d1789263/.github/workflows/test-and-deploy.yml#L3-L8

2. What does this workflow do?
This workflow always runs tests. If the tests pass, then it depends on the event that triggered the workflow. If the event was a merge to `master` branch, we deploy to github pages:
https://github.com/cardstack/boxel/blob/fe7b727ea0b2952eb364537b2236a2a8d1789263/.github/workflows/test-and-deploy.yml#L194-L198
If the event was an opening, reopening, or update of a PR to `master`, then we deploy a preview site to s3:
https://github.com/cardstack/boxel/blob/fe7b727ea0b2952eb364537b2236a2a8d1789263/.github/workflows/test-and-deploy.yml#L229-L232
If the event that triggered the workflow was an opening or reopening of a PR, regardless of whether the deployment to s3 worked, we add a comment to the PR with a link to the preview site. This makes sure that the link is near the top.
https://github.com/cardstack/boxel/blob/fe7b727ea0b2952eb364537b2236a2a8d1789263/.github/workflows/test-and-deploy.yml#L263-L270

3. How does the preview deployment work?
We're using [ember-cli-deploy](https://github.com/ember-cli-deploy/ember-cli-deploy) for it, which has a nice plugin system that includes deployment of assets and `index.html` to AWS S3. We tell `ember-cli-deploy` what plugins to use, and with what arguments, in [this file](https://github.com/cardstack/boxel/blob/preview-deployments/config/deploy.js).

File: ember-cli-build.js [hunk 1]
```diff
@@ -34,6 +34,10 @@ module.exports = function (defaults) {
     // staticComponents: true,
     staticAppPaths: ['data'],
     packagerOptions: {
+      publicAssetURL:
+        process.env.DEPLOY_TARGET === 's3-preview'
+          ? process.env.S3_PREVIEW_ASSET_BUCKET_ENDPOINT
```
Comment (lukemelia) [inline ember-cli-build.js]: I think this should have a "/" at the end of it

Comment (Aierie) [reply ember-cli-build.js]: Good catch, thanks!

**Comment by Aierie** [comment]
Deployment was successful! One issue with the current setup - with virtual-hosting-style access (https://boxel-preview-assets.cardstack.com.s3.amazonaws.com) , Boxel assets have to be accessed via http, not https. https has certificate issues because [you cannot have a '.' in the bucket name if you want to use https](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html). I changed our url to path-style for now (https://s3.us-east-1.amazonaws.com/boxel-preview-assets.cardstack.com), though AWS will be deprecating this at some point.

**Comment by Aierie** [comment]
Also, when loading the site earlier today, I found that load times for JS could go between 5 - 15 minutes. I think this is a region issue, but right now it's loading quickly again (whole site loads within 5 seconds on hard refresh) so I'm not too sure.

**Comment by lukemelia** [comment]
>Also, when loading the site earlier today, I found that load times for JS could go between 5 - 15 minutes.

Minutes, or seconds?

We could put cloudfront distributions in front of each bucket to help with this and the https issue.

**Comment by Aierie** [comment]
Actually took minutes to load a 2.1MB js file. Malaysian internet isn't great, but waiting for that long was a first for me. Loading was okay (comparable to gh pages) for @burieberry though, and it got a lot better for me later.

In terms of an MVP I think this might be enough to demo to the design team. We could merge this first and then aim to add Cloudfront distributions later?

**Comment by lukemelia** [comment]
>In terms of an MVP I think this might be enough to demo to the design team. We could merge this first and then aim to add Cloudfront distributions later?

Agreed.

**Review by lukemelia — APPROVED** [review]

**Pull request merged**